src/extension/type_scheme.rs

```diff
@@ -0,0 +1,269 @@
+    use crate::std_extensions::collections::{EXTENSION, LIST_TYPENAME};
+    use crate::types::type_param::{TypeArg, TypeArgError, TypeParam};
+    use crate::types::{CustomType, FunctionType, Type, TypeBound};
+    use crate::Extension;
+
+    use super::OpDefTypeScheme;
+
+    #[test]
+    fn test_opaque() -> Result<(), SignatureError> {
+        let list_def = EXTENSION.get_type(&LIST_TYPENAME).unwrap();
+        let tyvar = TypeArg::use_var(0, TypeParam::Type(TypeBound::Any));
+        let list_of_var = Type::new_extension(list_def.instantiate_concrete([tyvar.clone()])?);
+        let reg: ExtensionRegistry = [PRELUDE.to_owned(), EXTENSION.to_owned()].into();
+        let list_len = OpDefTypeScheme::new(
+            [TypeParam::Type(TypeBound::Any)],
+            FunctionType::new(vec![list_of_var], vec![USIZE_T]),
+            &reg,
+        )?;
+
+        let t = list_len.compute_signature(&[TypeArg::Type { ty: USIZE_T }], &reg)?;
+        assert_eq!(
+            t,
+            FunctionType::new(
+                vec![Type::new_extension(
+                    list_def
+                        .instantiate_concrete([TypeArg::Type { ty: USIZE_T }])
+                        .unwrap()
+                )],
+                vec![USIZE_T]
+            )
+        );
+
+        Ok(())
+    }
+
+    fn id_fn(t: Type) -> FunctionType {
+        FunctionType::new(vec![t.clone()], vec![t])
+    }
+
+    #[test]
+    fn test_mismatched_args() -> Result<(), SignatureError> {
+        let ar_def = PRELUDE.get_type("array").unwrap();
+        let typarams = [TypeParam::Type(TypeBound::Any), TypeParam::max_nat()];
+        let tyvar = TypeArg::use_var(0, typarams[0].clone());
+        let szvar = TypeArg::use_var(1, typarams[1].clone());
+
+        // Valid schema...
+        let good_array =
+            Type::new_extension(ar_def.instantiate_concrete([tyvar.clone(), szvar.clone()])?);
+        let good_ts = OpDefTypeScheme::new(typarams.clone(), id_fn(good_array), &PRELUDE_REGISTRY)?;
+
+        // Sanity check (good args)
+        good_ts.compute_signature(
+            &[TypeArg::Type { ty: USIZE_T }, TypeArg::BoundedNat { n: 5 }],
+            &PRELUDE_REGISTRY,
+        )?;
+
+        let wrong_args = good_ts.compute_signature(
+            &[TypeArg::BoundedNat { n: 5 }, TypeArg::Type { ty: USIZE_T }],
+            &PRELUDE_REGISTRY,
+        );
+        assert_eq!(
+            wrong_args,
+            Err(SignatureError::TypeArgMismatch(
+                TypeArgError::TypeMismatch {
+                    param: typarams[0].clone(),
+                    arg: TypeArg::BoundedNat { n: 5 }
+                }
+            ))
+        );
+
+        // (Try to) make a schema with bad args
+        let arg_err = SignatureError::TypeArgMismatch(TypeArgError::TypeMismatch {
+            param: typarams[0].clone(),
+            arg: szvar.clone(),
+        });
+        assert_eq!(
+            ar_def.instantiate_concrete([szvar.clone(), tyvar.clone()]),
```

Comment:
I find it a bit disturbing that the collections extension seems to have its own distinct notion of type variables

Reply:
Hmmm, I'm not sure what you mean here?

Quite a few extensions use custom binary functions to do things that this type-scheme system will do for them, so a follow-up PR is definitely to remove all that binary code and use type schemes. However, I don't think that needs to be part of this PR.